File: client/pose/shoulderPress.py
from pose.exerciseBase import *


class ShoulderPress(Exercise):
    def __init__(self, reps=10):
        super().__init__(reps, "Shoulder press")

    def is_start(self, landmarks):
        angle_left = calculate_angle(landmarks[landmark_index["left_shoulder"]],
                                     landmarks[landmark_index["left_elbow"]],
                                     landmarks[landmark_index["left_wrist"]])

        angle_right = calculate_angle(landmarks[landmark_index["right_shoulder"]],
                                      landmarks[landmark_index["right_elbow"]],
                                      landmarks[landmark_index["right_wrist"]])

        # return angle_right < 80 and angle_right < 80

        return angle_left <= 80 and angle_right <= 80 and landmarks[
            landmark_index["right_wrist"]].y < landmarks[landmark_index["right_shoulder"]].y and landmarks[
                   landmark_index["left_wrist"]].y < landmarks[landmark_index["left_shoulder"]].y

    def is_end(self, landmarks):
        angle_left = calculate_angle(landmarks[landmark_index["left_shoulder"]],
                                     landmarks[landmark_index["left_elbow"]],
                                     landmarks[landmark_index["left_wrist"]])

        angle_right = calculate_angle(landmarks[landmark_index["right_shoulder"]],
                                      landmarks[landmark_index["right_elbow"]],
                                      landmarks[landmark_index["right_wrist"]])

        # print(self.compute_percentage_done(angle_left))

        # return angle_left > 150 and angle_right > 150

        return angle_left >= 165 and angle_right >= 165 and landmarks[
            landmark_index["right_elbow"]].y < landmarks[landmark_index["right_shoulder"]].y and landmarks[
                   landmark_index["left_elbow"]].y < landmarks[landmark_index["left_shoulder"]].y

    def compute_percentage_done(self, angle):
        return np.interp(angle, (80, 160), (0, 100))

    def check_errors(self, landmarks):
        if not self.check_extension(landmarks):
            self.error = True
            return 0, "Try to keep your wrists(hands) right above the elbows"
        return 1, "You are doing good, keep it up"

    def check_extension(self, landmarks):
        return self.get_dist_x(landmarks[landmark_index["right_wrist"]],
                               landmarks[landmark_index["right_elbow"]]) < 0.15 \
               and self.get_dist_x(landmarks[landmark_index["left_wrist"]],
                                   landmarks[landmark_index["left_elbow"]]) < 0.15

    def get_dist_x(self, l1, l2):
        return abs(l1.x - l2.x)

File: client/Bachelor-Client/Bachelor-Client/Pose/ShoulderPress.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Bachelor_Client.Utils;
using Microsoft.Kinect;

namespace Bachelor_Client.Pose
{
    public class ShoulderPress : Exercise
    {
        public ShoulderPress(int reps) : base(reps)
        {
        }

        public override bool IsStart(JointCollection joints)
        {
            var angleLeft = JointsOperations.Get2DAngle(joints[JointType.ShoulderLeft],
                joints[JointType.ElbowLeft],
                joints[JointType.WristLeft]);

            var angleRight = JointsOperations.Get2DAngle(joints[JointType.ShoulderRight],
                joints[JointType.ElbowRight],
                joints[JointType.WristRight]);

            return angleLeft <= 100
                   && angleRight <= 100
                   && joints[JointType.WristRight].Position.Y > joints[JointType.ShoulderRight].Position.Y
                   && joints[JointType.WristLeft].Position.Y > joints[JointType.ShoulderLeft].Position.Y;
        }

        public override bool IsEnd(JointCollection joints)
        {
            var angleLeft = JointsOperations.Get2DAngle(joints[JointType.ShoulderLeft],
                joints[JointType.ElbowLeft],
                joints[JointType.WristLeft]);

            var angleRight = JointsOperations.Get2DAngle(joints[JointType.ShoulderRight],
                joints[JointType.ElbowRight],
                joints[JointType.WristRight]);

            return angleLeft >= 160
                   && angleRight >= 160
                   && joints[JointType.WristRight].Position.Y > joints[JointType.ShoulderRight].Position.Y
                   && joints[JointType.WristLeft].Position.Y > joints[JointType.ShoulderLeft].Position.Y;
        }

        public override KeyValuePair<int, string> CheckErrors(JointCollection joints)
        {
            if (!CheckExtension(joints))
            {
                isError = true;
                return new KeyValuePair<int, string>(0, "Try to keep your wrists(hands) right above the elbows");
            }
            return new KeyValuePair<int, string>(1, "You are doing good, keep it up");
        }

        private bool CheckExtension(JointCollection joints)
        {
            return GetDistanceX(joints[JointType.WristRight], joints[JointType.ElbowRight]) < 0.1f
                   && GetDistanceX(joints[JointType.WristLeft], joints[JointType.ElbowLeft]) < 0.1f;
        }
    }
}
